src/matvis/cpu/cpu.py

```diff
@@ -0,0 +1,248 @@
+"""CPU-based implementation of the matvis visibility simulator."""
+
+from __future__ import annotations
+
+import logging
+import numpy as np
+import psutil
+import time
+import tracemalloc as tm
+from astropy.coordinates import EarthLocation, SkyCoord
+from astropy.time import Time
+from collections.abc import Sequence
+
+# from pympler import tracker
+from pyuvdata import UVBeam
+from typing import Callable
+
+from .._utils import get_desired_chunks, get_dtypes, log_progress, logdebug, memtrace
+from ..core import _validate_inputs
+from ..core.coords import CoordinateRotation
+from ..core.getz import ZMatrixCalc
+from ..core.tau import TauCalculator
+from . import matprod as mp
+from .beams import UVBeamInterpolator
+
+logger = logging.getLogger(__name__)
+
+
+def simulate(
+    *,
+    antpos: np.ndarray,
+    freq: float,
+    times: Time,
+    skycoords: SkyCoord,
+    telescope_loc: EarthLocation,
+    I_sky: np.ndarray,
+    beam_list: Sequence[UVBeam | Callable] | None,
+    antpairs: np.ndarray | list[tuple[int, int]] | None = None,
+    precision: int = 1,
+    polarized: bool = False,
+    beam_idx: np.ndarray | None = None,
+    beam_spline_opts: dict | None = None,
+    max_progress_reports: int = 100,
+    matprod_method: str = "CPUMatMul",
```

Comment:
Probably more clear to write `Literal["CPUMatMul", "CPUVectorLoop"]`

Reply:
Good suggestion. I've done that.